File: pippip-native-ios/Support/NewAccountCreator.m
//
//  NewAccountCreator.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 10/10/17.
//  Copyright © 2017 seComm. All rights reserved.
//

#import "pippip_native_ios-Swift.h"

#import "NewAccountCreator.h"
#import "NewAccountRequest.h"
#import "NewAccountResponse.h"
#import "NewAccountFinish.h"
#import "NewAccountFinal.h"
#import "ApplicationSingleton.h"
#import "AccountSession.h"
#import "Notifications.h"
#import "AccountManager.h"
#import "CKPEMCodec.h"

typedef enum STEP { REQUEST, FINISH } ProcessStep;

@interface NewAccountCreator ()
{
    ProcessStep step;
    SessionState *sessionState;
    NSString *passphrase;
}

@property (weak, nonatomic) RESTSession *session;

@end

@implementation NewAccountCreator

@synthesize errorDelegate;
@synthesize postPacket;

- (instancetype)init {
    self = [super init];

    errorDelegate = [[NotificationErrorDelegate alloc] init:@"New Account Error"];
    _session = [ApplicationSingleton instance].restSession;
    sessionState = [[SessionState alloc] init];

    return self;

}

- (void)accountCreated {

    [self storeVault];
    sessionState.authenticated = true;
    AccountManager *accountManager = [[AccountManager alloc] init];
    [accountManager setDefaultConfig];

    [[NSNotificationCenter defaultCenter] postNotificationName:AUTHENTICATED object:sessionState];
    
}

- (void)createAccount:(NSString*)accountName withPassphrase:(NSString *)pass {

    [AccountManager accountName:accountName];
    passphrase = pass;
    ParameterGenerator *generator = [[ParameterGenerator alloc] init];
    [generator generateParameters:accountName];
    NSMutableDictionary *info = [NSMutableDictionary dictionary];
    info[@"progress"] = [NSNumber numberWithFloat:0.25];
    [[NSNotificationCenter defaultCenter] postNotificationName:@"UpdateProgress" object:nil userInfo:info];
    // Start the session.
    [_session startSession:self];

}

- (void) doFinish {

    step = FINISH;
    postPacket = [[NewAccountFinish alloc] init];
    [_session queuePost:self];

}

// This will be called after the new account request completes
- (void)postComplete:(NSDictionary*)response {

    if (response != nil) {
        switch (step) {
            case REQUEST:
                if ([self validateResponse:response]) {
                    NSMutableDictionary *info = [NSMutableDictionary dictionary];
                    info[@"progress"] = [NSNumber numberWithFloat:0.75];
                    [[NSNotificationCenter defaultCenter] postNotificationName:@"UpdateProgress"
                                                                        object:nil userInfo:info];
                    [self doFinish];
                }
                break;
            case FINISH:
                if ([self validateFinish:response]) {
                    NSMutableDictionary *info = [NSMutableDictionary dictionary];
                    info[@"progress"] = [NSNumber numberWithFloat:1.0];
                    [[NSNotificationCenter defaultCenter] postNotificationName:@"UpdateProgress"
                                                                        object:nil userInfo:info];
                    [self accountCreated];
                }
                break;
        }

    }

}

- (void)sessionComplete:(NSDictionary*)response {

    if (response != nil) {
        NSString *sessionId = [response objectForKey:@"sessionId"];
        NSString *serverPublicKeyPEM = [response objectForKey:@"serverPublicKey"];
        if (sessionId == nil) {
            [errorDelegate sessionError:@"Invalid server response, missing session ID"];
        }
        else if (serverPublicKeyPEM == nil) {
            [errorDelegate sessionError:@"Invalid server response, missing public key"];
        }
        else {
            NSMutableDictionary *info = [NSMutableDictionary dictionary];
            info[@"progress"] = [NSNumber numberWithFloat:0.5];
            [[NSNotificationCenter defaultCenter] postNotificationName:@"UpdateProgress" object:nil userInfo:info];
            sessionState.sessionId = [sessionId intValue];
            CKPEMCodec *pem = [[CKPEMCodec alloc] init];
            sessionState.serverPublicKey = [pem decodePublicKey:serverPublicKeyPEM];
            step = REQUEST;
            postPacket = [[NewAccountRequest alloc] init];
            [_session queuePost:self];
        }
    }

}

- (void)storeVault {
    
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *docPath = [paths objectAtIndex:0];
    NSString *vaultsPath = [docPath stringByAppendingPathComponent:@"PippipVaults"];
    NSString *vaultPath = [vaultsPath stringByAppendingPathComponent:[AccountManager accountName]];
  
    UserVault *vault = [[UserVault alloc] init];
    NSError *error = nil;
    NSData *vaultData = [vault encode:passphrase error:&error];

    if (error == nil) {
        NSFileManager *fileManager = [NSFileManager defaultManager];
        [fileManager createDirectoryAtPath:vaultsPath
               withIntermediateDirectories:NO
                                attributes:nil
                                     error:nil];
        [vaultData writeToFile:vaultPath atomically:YES];
    }
    else {
        NSLog(@"Error storing vault = %@", error);
    }

}

- (BOOL) validateFinish:(NSDictionary*)response {

    NewAccountFinal *accountFinal = [[NewAccountFinal alloc] init];
    return [accountFinal processResponse:response
                           errorDelegate:errorDelegate];

}

- (BOOL) validateResponse:(NSDictionary*)response {

    NewAccountResponse *accountResponse = [[NewAccountResponse alloc] init];
    return [accountResponse processResponse:response
                              errorDelegate:errorDelegate];

}

@end

File: pippip-native-ios/Support/NewAccountCreator.swift
//
//  NewAccountCreator.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 5/7/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit

enum Step { case finish, none, request }

class NewAccountCreator: NSObject, RequestProcessProtocol {

    var postPacket: PostPacket?
    var errorDelegate: ErrorDelegate
    var restSession: RESTSession
    var sessionState = SessionState()
    var passphrase = ""
    var step = Step.none

    override init() {

        errorDelegate = NotificationErrorDelegate("New Account Error")
        restSession = ApplicationSingleton.instance().restSession

    }

    func accountCreated() {

        do {
            try storeVault()
            sessionState.authenticated = true
            let accountManager = AccountManager()
            accountManager.setDefaultConfig()
            NotificationCenter.default.post(name: Notifications.Authenticated, object: nil)
        }
        catch {
            var info = [AnyHashable: Any]()
            info["title"] = "New Account Error"
            info["message"] = "Error storing user vault"
            NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
            print("Error storing user vault: \(error)")
        }

    }

    func createAccount(accountName: String, passphrase: String) {

        AccountManager.accountName = accountName
        self.passphrase = passphrase
        let parameterGenerator = ParameterGenerator()
        parameterGenerator.generateParameters(accountName)
        var info = [AnyHashable: Any]()
        info["progress"] = Float(0.25)
        NotificationCenter.default.post(name: Notifications.UpdateProgress, object: nil, userInfo: info)
        restSession.start(self)
        
    }

    func doFinish() {

        step = Step.finish
        postPacket = NewAccountFinish()
        restSession.queuePost(self)

    }

    func postComplete(_ response: [AnyHashable : Any]?) {

        if (response != nil) {
            switch (step) {
            case Step.request:
                if validateResponse(response!) {
                    var info = [AnyHashable: Any]()
                    info["progress"] = Float(0.25)
                    NotificationCenter.default.post(name: Notifications.UpdateProgress, object: nil, userInfo: info)
                    doFinish();
                }
                break;
            case Step.finish:
                if validateFinish(response!) {
                    var info = [AnyHashable: Any]()
                    info["progress"] = Float(0.25)
                    NotificationCenter.default.post(name: Notifications.UpdateProgress, object: nil, userInfo: info)
                    accountCreated();
                }
                break;
            case Step.none:
                break;
            }
            
        }
        
    }
    
    func sessionComplete(_ response: [AnyHashable : Any]?) {

        if (response != nil) {
            guard let sessionId = response!["sessionId"] as? Int32 else {
                errorDelegate.sessionError("Invalid server response, missing session ID")
                return
            }
            guard let serverPublicKeyPEM = response!["serverPublicKey"] as? String else {
                errorDelegate.sessionError("Invalid server response, missing public key")
                return
            }
            sessionState.sessionId = sessionId
            let pem = CKPEMCodec()
            sessionState.serverPublicKey = pem.decodePublicKey(serverPublicKeyPEM)
            step = Step.request
            postPacket = NewAccountRequest()
            restSession.queuePost(self)
        }
        
    }

    func storeVault() throws {

        let fileManager = FileManager.default
        let documentDirectory = try fileManager.url(for: .documentDirectory,
                                                    in: .userDomainMask,
                                                    appropriateFor:nil,
                                                    create:false)
        let vaultsURL = documentDirectory.appendingPathComponent("PippipVaults")
        let vaultURL = vaultsURL.appendingPathComponent(AccountManager.accountName!)
        
        let vault = UserVault()
        let vaultData = try vault.encode(passphrase)
        try vaultData.write(to: vaultURL)

    }

    func validateFinish(_ response: [AnyHashable: Any]) -> Bool {

        let accountFinal = NewAccountFinal()
        return accountFinal.processResponse(response, errorDelegate: errorDelegate)

    }
    
    func validateResponse(_ response: [AnyHashable: Any]) -> Bool {

        let accountResponse = NewAccountResponse()
        return accountResponse.processResponse(response, errorDelegate: errorDelegate)

    }

}
